### ISSUE TITLE
Unable to retrieve logs using kubectl and Linode

### ISSUE THREAD

**Issue by DazWilkin** [deislabs/krustlet#492]
I'm able to run Krustlet against LKE using both WASI|WASCC and apply Pods but, I'm unable to retrieve logs using `kubectl logs`. I am able to retrieve the logs if `curl` the equivalent host URL!?

At first, I thought the issue was related to #168 but Linode Kubernetes Engine (LKE) appears to address the Krustlet by IP not hostname.

I've successfully deployed Krustlet w/ other managed Kubernetes offerings and repeating my steps with LKE.

I'm using Bootstrap Tokens and switching between WASI and WASCC.

But:

VM:
```bash
IP="192.53.124.117" # Will delete the Linode

KUBECONFIG=${PWD}/krustlet.config \
./krustlet-wascc \
--node-ip=${IP} \
--node-name=krustlet \
--bootstrap-file=$PWD/bootstrap.conf \
--cert-file=${PWD}/krustlet.crt \
--private-key-file=${PWD}/krustlet.key
```
And:
```bash
kubectl certificate approve localhost-tls
certificatesigningrequest.certificates.k8s.io/localhost-tls approved

kubectl apply \
--filename=${DEMOS}/wascc/hello-world-assemblyscript/k8s.yaml

kubectl logs pod/hello-world-wascc-assemblyscript
Error from server: Get https://192.53.124.117:3000/containerLogs/default/hello-world-wascc-assemblyscript/hello-world-wascc-assemblyscript: dial tcp 192.53.124.117:3000: i/o timeout

curl \
--insecure \
--write-out '%{response_code}' \
https://192.53.124.117:3000/containerLogs/default/hello-world-wascc-assemblyscript/hello-world-wascc-assemblyscript
200
```

Restarting with WASI:

```bash
kubectl delete node/krustlet

kubectl apply --filename="${DEMOS}/wasi/hello-world-rust/k8s.yaml"

kubectl logs pod/hello-world-wasi-rust
Error from server: Get https://192.53.124.117:3000/containerLogs/default/hello-world-wasi-rust/hello-world-wasi-rust: dial tcp 192.53.124.117:3000: i/o timeout

curl \
--insecure \
https://192.53.124.117:3000/containerLogs/default/hello-world-wasi-rust/hello-world-wasi-rust
hello from stdout!
hello from stderr!
CONFIG_MAP_VAL=cool stuff
POD_NAME=hello-world-wasi-rust
FOO=bar
Args are: []

Bacon ipsum dolor amet chuck turducken porchetta, tri-tip spare ribs t-bone ham hock. Meatloaf
pork belly leberkas, ham beef pig corned beef boudin ground round meatball alcatra jerky.
Pancetta brisket pastrami, flank pork chop ball tip short loin burgdoggen. Tri-tip kevin
shoulder cow andouille. Prosciutto chislic cupim, short ribs venison jerky beef ribs ham hock
short loin fatback. Bresaola meatloaf capicola pancetta, prosciutto chicken landjaeger andouille
swine kielbasa drumstick cupim tenderloin chuck shank. Flank jowl leberkas turducken ham tongue
beef ribs shankle meatloaf drumstick pork t-bone frankfurter tri-tip.
```

Here's the node:

```bash
kubectl describe node/krustlet
Name:               krustlet
Roles:              <none>
Labels:             beta.kubernetes.io/arch=wasm32-wasi
                    beta.kubernetes.io/os=linux
                    kubernetes.io/arch=wasm32-wasi
                    kubernetes.io/hostname=localhost
                    kubernetes.io/os=linux
                    type=krustlet
Annotations:        node.alpha.kubernetes.io/ttl: 0
                    volumes.kubernetes.io/controller-managed-attach-detach: true
CreationTimestamp:  Mon, 25 Jan 2021 14:16:14 -0800
Taints:             kubernetes.io/arch=wasm32-wasi:NoExecute
Unschedulable:      false
Lease:
  HolderIdentity:  krustlet
  AcquireTime:     Mon, 25 Jan 2021 14:20:15 -0800
  RenewTime:       Mon, 25 Jan 2021 14:20:15 -0800
Conditions:
  Type    Status  LastHeartbeatTime                 LastTransitionTime                Reason         Message
  ----    ------  -----------------                 ------------------                ------         -------
  Ready   True    Mon, 25 Jan 2021 14:20:15 -0800   Mon, 01 Jan 0001 00:00:00 +0000   KubeletReady   kubelet is posting ready status
Addresses:
  InternalIP:  192.53.124.117
  Hostname:    localhost
Capacity:
  cpu:                4
  ephemeral-storage:  61255492Ki
  hugepages-1Gi:      0
  hugepages-2Mi:      0
  memory:             4032800Ki
  pods:               110
Allocatable:
  cpu:                4
  ephemeral-storage:  61255492Ki
  hugepages-1Gi:      0
  hugepages-2Mi:      0
  memory:             4032800Ki
  pods:               110
System Info:
  Machine ID:                 
  System UUID:                
  Boot ID:                    
  Kernel Version:             
  OS Image:                   
  Operating System:           linux
  Architecture:               wasm-wasi
  Container Runtime Version:  mvp
  Kubelet Version:            0.5.0
  Kube-Proxy Version:         v1.17.0
PodCIDR:                      10.244.0.0/24
PodCIDRs:                     10.244.0.0/24
Non-terminated Pods:          (2 in total)
  Namespace                   Name                 CPU Requests  CPU Limits  Memory Requests  Memory Limits  AGE
  ---------                   ----                 ------------  ----------  ---------------  -------------  ---
  kube-system                 calico-node-tgrnm    250m (6%)     0 (0%)      0 (0%)           0 (0%)         4m1s
  kube-system                 kube-proxy-2cm7k     0 (0%)        0 (0%)      0 (0%)           0 (0%)         4m1s
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests   Limits
  --------           --------   ------
  cpu                250m (6%)  0 (0%)
  memory             0 (0%)     0 (0%)
  ephemeral-storage  0 (0%)     0 (0%)
  hugepages-1Gi      0 (0%)     0 (0%)
  hugepages-2Mi      0 (0%)     0 (0%)
Events:              <none>
```

> **NOTE** Linodes aren't provisioned with host names and don't have DNS entries.

```bash
more /etc/hosts
127.0.0.1	localhost

# The following lines are desirable for IPv6 capable hosts
::1     localhost ip6-localhost ip6-loopback
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
```

**Comment by bacongobbler** [comment]
That error indicates that the whatever node the Kubernetes API server is running on cannot make a connection to the Linode VM, but your local machine can. Which to me indicates that the firewall configuration for the linode VM is misconfigured.

In general you should not be able to reach the krustlet API server directly from your laptop - you should only be able to reach the API through the VPN (or through `kubectl logs`) to reach the node.

I would check and ensure two things:

1. Add the Linode VM to the LKE instance's VPN.
1. Disallow all inbound connections from outside the firewall. Only allow connections from within the VPN.

**Comment by bacongobbler** [comment]
You can have a look at the Azure documentation to see how this is accomplished: https://github.com/deislabs/krustlet/blob/9a6182276b1835daf641f0007f27b50dfd55343b/contrib/azure/azuredeploy.json#L202-L208

**Comment by DazWilkin** [comment]
Thanks @bacongobbler 

This isn't a blocking issue. This is just a dev scenario and guidance to run Krustlet against Linode Kubernetes Engine clusters.

It's possible that the API Server is unable to access the Krustlet.

I did not create a firewall when provisioning the VM and I am able to access the Krustlet from a node:

```bash
curl --insecure https://45.79.102.137:3000/containerLogs/default/hello-world-wasi-rust/hello-world-wasi-rust
hello from stdout!
hello from stderr!
CONFIG_MAP_VAL=cool stuff
POD_NAME=hello-world-wasi-rust
FOO=bar
Args are: []

Bacon ipsum dolor amet chuck turducken porchetta, tri-tip spare ribs t-bone ham hock. Meatloaf
pork belly leberkas, ham beef pig corned beef boudin ground round meatball alcatra jerky.
Pancetta brisket pastrami, flank pork chop ball tip short loin burgdoggen. Tri-tip kevin
shoulder cow andouille. Prosciutto chislic cupim, short ribs venison jerky beef ribs ham hock
short loin fatback. Bresaola meatloaf capicola pancetta, prosciutto chicken landjaeger andouille
swine kielbasa drumstick cupim tenderloin chuck shank. Flank jowl leberkas turducken ham tongue
beef ribs shankle meatloaf drumstick pork t-bone frankfurter tri-tip.
```

I was wondering whether the issue is to do with the Linode being addressed solely by IP and without a hostname (defaults to `localhost`).

Here's the Krustlet's cert:

```bash
openssl x509 -in linode.crt -text -noout
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            5f:f8:f1:6d:...
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = kubernetes
        Validity
            Not Before: Jan 27 17:19:00 2021 GMT
            Not After : Jan 27 17:19:00 2022 GMT
        Subject: O = system:nodes, CN = system:node:localhost
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:77:8b:fa:...
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Alternative Name: 
                DNS:localhost, IP Address:45.79.102.137
    Signature Algorithm: sha256WithRSAEncryption
         4c:a3:d3:10:...
```

**Comment by bacongobbler** [comment]
Oh! I didn't realize we still haven't merged #497 yet. It looks like another maintainer caught this as well during PR review, so we can track that discussion as part of the review.

Let's follow up and resolve this issue in the PR before we start opening tickets. Thanks!
